In [164]:
%load_ext autoreload
%autoreload 2

The autoreload extension is already loaded. To reload it, use:
  %reload_ext autoreload


In [165]:
import pandas as pd
from functions import main_cleaning

pd.set_option('display.max_columns', None) 

df_raw = pd.read_csv("dielectron.csv")

In [166]:
df_cleaned = main_cleaning(df_raw)

e:\Iván\Documentos\Clase\Informatica\Beca IronHack\Temario\PROYECTOS\Proyecto_final_1710\functions.py:12: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df["is_same_charge"] = df["Q1"] == df["Q2"]
e:\Iván\Documentos\Clase\Informatica\Beca IronHack\Temario\PROYECTOS\Proyecto_final_1710\functions.py:19: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df["is_outlier"] = False


In [167]:
df_cleaned["Event"].nunique()

99976

# MACHINE LEARNING

### Los resultados de MSE, MAE y R2 para KNN sin normalizar eran: 
MAE: 17.717622715597013  
MSE: 12.581512831873468   
R2: 0.507506217561434  
### So I will normalize/standarize.

In [168]:
df_ml = df_cleaned.dropna()

In [267]:
df_ml

Run      Event        E1        px1        py1       pz1       pt1  \
0      147115  366639895  58.71410  -7.311320  10.531000 -57.29740  12.82020   
1      147115  366704169   6.61188  -4.152130  -0.579855  -5.11278   4.19242   
2      147115  367112316  25.54190 -11.480900   2.041680  22.72460  11.66100   
3      147115  366952149  65.39590   7.512140  11.887100  63.86620  14.06190   
4      147115  366523212  61.45040   2.952840 -14.622700 -59.61210  14.91790   
...       ...        ...       ...        ...        ...       ...       ...   
99995  146511  522575834  12.31310 -10.658000   5.164440   3.36858  11.84330   
99996  146511  522786431  18.46420   7.854990  15.133000  -7.08659  17.05020   
99997  146511  522906124   4.18566  -3.273500  -0.308507  -2.59013   3.28801   
99998  146511  523243830  54.46220  11.352600  11.880900  51.92400  16.43280   
99999  146511  524172389   7.64000   0.886162   5.478900  -5.25033   5.55010   

           eta1      phi1  Q1        E2        px2      py2        pz2  \
0     -2.202670  2.177660   1  11.28360  -1.032340 -1.88066 -11.077800   
1     -1.028420 -3.002840  -1  17.14920 -11.713500  5.04474  11.464700   
2      1.420480  2.965600   1  15.82030  -1.472800  2.25895 -15.588800   
3      2.218380  1.007210   1  25.12730   4.087860  2.59641  24.656300   
4     -2.093750 -1.371540  -1  13.88710  -0.277757 -2.42560 -13.670800   
...         ...       ...  ..       ...        ...      ...        ...   
99995  0.280727  2.690370  -1   1.80181   0.668609 -1.58437   0.537805   
99996 -0.404510  1.092010   1  14.69110  -1.418020 -2.28117 -14.443500   
99997 -0.723075 -3.047630   1  72.81740 -11.074900 -9.28179 -71.369300   
99998  1.867800  0.808132  -1   8.58671   0.378009  3.07828   8.007050   
99999 -0.842662  1.410440   1  52.10880  16.807500 -4.60510  49.108400   

            pt2      eta2      phi2  Q2         M  is_same_charge  is_outlier  
0       2.14537 -2.344030 -2.072810  -1   8.94841           False       False  
1      12.75360  0.808077  2.734920   1  15.89300           False       False  
2       2.69667 -2.455080  2.148570   1  38.38770            True       False  
3       4.84272  2.330210  0.565865  -1   3.72862           False       False  
4       2.44145 -2.423700 -1.684810  -1   2.74718            True       False  
...         ...       ...       ...  ..       ...             ...         ...  
99995   1.71967  0.307851 -1.171470   1   8.44779           False       False  
99996   2.68598 -2.383880 -2.126960   1  20.71540            True       False  
99997  14.45010 -2.300410 -2.444050  -1  12.71350           False       False  
99998   3.10141  1.677170  1.448610   1   4.69670           False       False  
99999  17.42700  1.759250 -0.267427  -1  36.50430           False       False  

[99892 rows x 21 columns]

In [170]:
# features = df_ml[["Run", "E1", "eta1", "phi1", "E2", "eta2", "phi2", "is_same_charge", "is_outlier"]]
# target = df_ml[["M"]]

# Worst results

In [171]:
features = df_ml[["Run", "px1", "py1", "pz1", "px2", "py2", "pz2", "is_same_charge", "is_outlier"]]
target = df_ml[["M"]]

In [172]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(features, target, test_size=0.20, random_state=42)

### Normalization

In [173]:
from sklearn.preprocessing import MinMaxScaler

normalizer = MinMaxScaler()

X_train_norm= normalizer.fit_transform(X_train)        

X_test_norm = normalizer.transform(X_test)

In [174]:
X_train_norm = pd.DataFrame(X_train_norm, columns = X_train.columns)   
X_test_norm = pd.DataFrame(X_test_norm, columns = X_test.columns)

### I will try 5 different models

In [175]:
from sklearn.ensemble import RandomForestRegressor
from sklearn.preprocessing import StandardScaler
from sklearn.neighbors import KNeighborsRegressor
from sklearn.linear_model import LinearRegression
from sklearn.tree import DecisionTreeRegressor
from sklearn.svm import SVR
from sklearn.metrics import root_mean_squared_error, mean_absolute_error, r2_score

# Define models
models = {
  'KNeighborsRegressor': KNeighborsRegressor(),
  'Linear Regression': LinearRegression(),
  'Decision Tree': DecisionTreeRegressor(random_state=42),
  'Random Forest': RandomForestRegressor(random_state=42, n_estimators=100),
  'SVR': SVR(),
}   

# Train and evaluate models
results = {}

for name, model in models.items():
  # Train the model
    model.fit(X_train_norm, y_train)
  
  # Make predictions
    y_pred_norm = model.predict(X_test_norm)

    mae_standardized = mean_absolute_error(y_test, y_pred_norm)
    mse_standardized = root_mean_squared_error(y_test, y_pred_norm)
    r2_standardized = r2_score(y_test, y_pred_norm)
  
  # Store results
    results[name] = {
        'mae' : mae_standardized,
        'mse' : mse_standardized,
        'R2' : r2_standardized
  }

# Print results
for name, result in results.items():
  print(f"\n{name}:")
  print(f"Mean Absolute Error: {result['mae']}")
  print(f"Root Mean Squared Error: {result['mse']}")
  print(f"R2: {result['R2']}")

# Compare R2_value
r2_values = {name: result['R2'] for name, result in results.items()}
best_model = max(r2_values, key=r2_values.get)

print("\nModel Accuracy Comparison:")
for name, accuracy in r2_values.items():
  print(f"{name}: {accuracy:.4f}")

print(f"\nBest performing model: {best_model} with accuracy {r2_values[best_model]:.4f}")

c:\Anaconda\Lib\site-packages\sklearn\base.py:1474: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return fit_method(estimator, *args, **kwargs)
c:\Anaconda\Lib\site-packages\sklearn\utils\validation.py:1300: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)



KNeighborsRegressor:
Mean Absolute Error: 10.12530578237149
Root Mean Squared Error: 15.381530427495692
R2: 0.628816253440426

Linear Regression:
Mean Absolute Error: 19.259107458476098
Root Mean Squared Error: 24.897594748274074
R2: 0.02746634175662188

Decision Tree:
Mean Absolute Error: 4.1532178397317185
Root Mean Squared Error: 7.146840026724198
R2: 0.9198657871470972

Random Forest:
Mean Absolute Error: 2.3478761082736876
Root Mean Squared Error: 3.918134937140667
R2: 0.9759148902055805

SVR:
Mean Absolute Error: 16.845790601770357
Root Mean Squared Error: 24.50329975751872
R2: 0.0580258193386155

Model Accuracy Comparison:
KNeighborsRegressor: 0.6288
Linear Regression: 0.0275
Decision Tree: 0.9199
Random Forest: 0.9759
SVR: 0.0580

Best performing model: Random Forest with accuracy 0.9759


### Standarization

In [176]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()

X_train_scaled = scaler.fit_transform(X_train)        

X_test_scaled = scaler.transform(X_test)

In [177]:
X_train_scaled = pd.DataFrame(X_train_scaled, columns = X_train.columns)   
X_test_scaled = pd.DataFrame(X_test_scaled, columns = X_test.columns)

In [178]:
from sklearn.ensemble import RandomForestRegressor
from sklearn.neighbors import KNeighborsRegressor
from sklearn.linear_model import LinearRegression
from sklearn.tree import DecisionTreeRegressor
from sklearn.svm import SVR
from sklearn.metrics import root_mean_squared_error, mean_absolute_error, r2_score

# Define models
models = {
  'KNeighborsRegressor': KNeighborsRegressor(),
  'Linear Regression': LinearRegression(),
  'Decision Tree': DecisionTreeRegressor(random_state=42),
  'Random Forest': RandomForestRegressor(random_state=42, n_estimators=100),
  'SVR': SVR(),
}   

# Train and evaluate models
results = {}

for name, model in models.items():
  # Train the model
    model.fit(X_train_scaled, y_train)
  
  # Make predictions
    y_pred_scaled = model.predict(X_test_scaled)

    mae_standardized = mean_absolute_error(y_test, y_pred_scaled)
    mse_standardized = root_mean_squared_error(y_test, y_pred_scaled)
    r2_standardized = r2_score(y_test, y_pred_scaled)
  
  # Store results
    results[name] = {
        'mae' : mae_standardized,
        'mse' : mse_standardized,
        'R2' : r2_standardized
  }

# Print results
for name, result in results.items():
  print(f"\n{name}:")
  print(f"Mean Absolute Error: {result['mae']}")
  print(f"Root Mean Squared Error: {result['mse']}")
  print(f"R2: {result['R2']}")

# Compare R2_value
r2_values = {name: result['R2'] for name, result in results.items()}
best_model = max(r2_values, key=r2_values.get)

print("\nModel Accuracy Comparison:")
for name, accuracy in r2_values.items():
  print(f"{name}: {accuracy:.4f}")

print(f"\nBest performing model: {best_model} with accuracy {r2_values[best_model]:.4f}")

c:\Anaconda\Lib\site-packages\sklearn\base.py:1474: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return fit_method(estimator, *args, **kwargs)
c:\Anaconda\Lib\site-packages\sklearn\utils\validation.py:1300: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)



KNeighborsRegressor:
Mean Absolute Error: 8.875525099854848
Root Mean Squared Error: 13.74404795685576
R2: 0.7036402792529388

Linear Regression:
Mean Absolute Error: 19.259107458476098
Root Mean Squared Error: 24.897594748274077
R2: 0.02746634175662166

Decision Tree:
Mean Absolute Error: 4.142311697782672
Root Mean Squared Error: 7.141685544711038
R2: 0.919981335111248

Random Forest:
Mean Absolute Error: 2.3447513779118077
Root Mean Squared Error: 3.9098505538185626
R2: 0.9760166321541155

SVR:
Mean Absolute Error: 9.141582787416484
Root Mean Squared Error: 13.338408631516149
R2: 0.7208755440655801

Model Accuracy Comparison:
KNeighborsRegressor: 0.7036
Linear Regression: 0.0275
Decision Tree: 0.9200
Random Forest: 0.9760
SVR: 0.7209

Best performing model: Random Forest with accuracy 0.9760


### We take Standarized Random Forest as the best model

### Now, we are going to try RF removing columns

In [270]:
features_lower = df_ml[["Run", "pz1", "pz2", "is_same_charge", "is_outlier"]]
target_lower = df_ml[["M"]]

In [271]:
from sklearn.model_selection import train_test_split

X_train_lower, X_test_lower, y_train_lower, y_test_lower = train_test_split(features_lower, target_lower, test_size=0.20, random_state=42)

In [272]:
scaler_lower = StandardScaler()

X_train_lower_scaled = scaler_lower.fit_transform(X_train_lower)        

X_test_lower_scaled = scaler_lower.transform(X_test_lower)

In [273]:
X_train_lower_scaled = pd.DataFrame(X_train_lower_scaled, columns = X_train_lower.columns)   
X_test_lower_scaled = pd.DataFrame(X_test_lower_scaled, columns = X_test_lower.columns)

In [274]:
rf_lower = RandomForestRegressor(random_state=42, n_estimators=100)

rf_lower.fit(X_train_lower_scaled, y_train_lower)
  
y_pred_lower_scaled = rf_lower.predict(X_test_lower_scaled)

mae_standardized = mean_absolute_error(y_test_lower, y_pred_lower_scaled)
mse_standardized = root_mean_squared_error(y_test_lower, y_pred_lower_scaled)
r2_standardized = r2_score(y_test_lower, y_pred_lower_scaled)

mae_standardized, mse_standardized, r2_standardized

c:\Anaconda\Lib\site-packages\sklearn\base.py:1474: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return fit_method(estimator, *args, **kwargs)


(9.648483653205867, 15.70953893783305, 0.6128165981972764)

Todo -> (2.3447513779118077, 3.9098505538185626, 0.9760166321541155)

Sin py -> (5.030465963967166, 8.80652660636964, 0.8783256160020667)

Sin py ni pz -> (16.016459623359527, 21.49041157214318, 0.2754317177361222)

Sin px ni py -> (9.648483653205867, 15.70953893783305, 0.6128165981972764)

In [275]:
features_scaled = scaler_lower.transform(features_lower)

y_pred_lower_real = rf_lower.predict(features_scaled)


mae_standardized = mean_absolute_error(target_lower, y_pred_lower_real)
mse_standardized = root_mean_squared_error(target_lower, y_pred_lower_real)
r2_standardized = r2_score(target_lower, y_pred_lower_real)

mae_standardized, mse_standardized, r2_standardized

c:\Anaconda\Lib\site-packages\sklearn\base.py:493: UserWarning: X does not have valid feature names, but RandomForestRegressor was fitted with feature names
  warnings.warn(


(4.804240913842951, 8.788852094326584, 0.8789006392171295)

In [278]:
df_ml["pred"] = y_pred_lower_real

df_ml

C:\Users\Iván\AppData\Local\Temp\ipykernel_32284\2496337083.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_ml["pred"] = y_pred_lower_real


Run      Event        E1        px1        py1       pz1       pt1  \
0      147115  366639895  58.71410  -7.311320  10.531000 -57.29740  12.82020   
1      147115  366704169   6.61188  -4.152130  -0.579855  -5.11278   4.19242   
2      147115  367112316  25.54190 -11.480900   2.041680  22.72460  11.66100   
3      147115  366952149  65.39590   7.512140  11.887100  63.86620  14.06190   
4      147115  366523212  61.45040   2.952840 -14.622700 -59.61210  14.91790   
...       ...        ...       ...        ...        ...       ...       ...   
99995  146511  522575834  12.31310 -10.658000   5.164440   3.36858  11.84330   
99996  146511  522786431  18.46420   7.854990  15.133000  -7.08659  17.05020   
99997  146511  522906124   4.18566  -3.273500  -0.308507  -2.59013   3.28801   
99998  146511  523243830  54.46220  11.352600  11.880900  51.92400  16.43280   
99999  146511  524172389   7.64000   0.886162   5.478900  -5.25033   5.55010   

           eta1      phi1  Q1        E2        px2      py2        pz2  \
0     -2.202670  2.177660   1  11.28360  -1.032340 -1.88066 -11.077800   
1     -1.028420 -3.002840  -1  17.14920 -11.713500  5.04474  11.464700   
2      1.420480  2.965600   1  15.82030  -1.472800  2.25895 -15.588800   
3      2.218380  1.007210   1  25.12730   4.087860  2.59641  24.656300   
4     -2.093750 -1.371540  -1  13.88710  -0.277757 -2.42560 -13.670800   
...         ...       ...  ..       ...        ...      ...        ...   
99995  0.280727  2.690370  -1   1.80181   0.668609 -1.58437   0.537805   
99996 -0.404510  1.092010   1  14.69110  -1.418020 -2.28117 -14.443500   
99997 -0.723075 -3.047630   1  72.81740 -11.074900 -9.28179 -71.369300   
99998  1.867800  0.808132  -1   8.58671   0.378009  3.07828   8.007050   
99999 -0.842662  1.410440   1  52.10880  16.807500 -4.60510  49.108400   

            pt2      eta2      phi2  Q2         M  is_same_charge  is_outlier  \
0       2.14537 -2.344030 -2.072810  -1   8.94841           False       False   
1      12.75360  0.808077  2.734920   1  15.89300           False       False   
2       2.69667 -2.455080  2.148570   1  38.38770            True       False   
3       4.84272  2.330210  0.565865  -1   3.72862           False       False   
4       2.44145 -2.423700 -1.684810  -1   2.74718            True       False   
...         ...       ...       ...  ..       ...             ...         ...   
99995   1.71967  0.307851 -1.171470   1   8.44779           False       False   
99996   2.68598 -2.383880 -2.126960   1  20.71540            True       False   
99997  14.45010 -2.300410 -2.444050  -1  12.71350           False       False   
99998   3.10141  1.677170  1.448610   1   4.69670           False       False   
99999  17.42700  1.759250 -0.267427  -1  36.50430           False       False   

            pred  
0       8.597872  
1      19.626943  
2      39.525085  
3       8.538839  
4       6.694673  
...          ...  
99995   9.816381  
99996  18.591266  
99997  15.340402  
99998   7.410890  
99999  45.550666  

[99892 rows x 22 columns]

In [277]:
from sklearn.model_selection import cross_val_score

# Aplicamos la validación cruzada en el conjunto de entrenamiento
scores = cross_val_score(rf_lower, X_train_lower_scaled, y_train_lower, cv=5, scoring='r2')

print("Cross-validated R2 scores:", scores)
print("Mean cross-validated R2:", scores.mean())


c:\Anaconda\Lib\site-packages\sklearn\base.py:1474: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return fit_method(estimator, *args, **kwargs)
c:\Anaconda\Lib\site-packages\sklearn\base.py:1474: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return fit_method(estimator, *args, **kwargs)
c:\Anaconda\Lib\site-packages\sklearn\base.py:1474: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return fit_method(estimator, *args, **kwargs)
c:\Anaconda\Lib\site-packages\sklearn\base.py:1474: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return fit_method(estimato

Cross-validated R2 scores: [0.62209734 0.61593845 0.61195166 0.59977886 0.59800937]
Mean cross-validated R2: 0.6095551354410731


Sin px -> Mean cross-validated R2: 0.8715048380715927

Sin px ni py -> Mean cross-validated R2: 0.6095551354410731